packages/contracts/contracts/bundle/TokenBundle.sol

```diff
@@ -0,0 +1,168 @@
+
+        require(targetCount > 0, "!Tokens");
+        require(bundle[bundleId_].count == 0, "Token bundle id exists");
+
+        for (uint256 i = 0; i < targetCount; i += 1) {
+            _checkTokenType(_tokensToBind[i]);
+            bundle[bundleId_].collaterals[i] = _tokensToBind[i];
+        }
+
+        bundle[bundleId_].count = targetCount;
+    }
+
+    /// @dev Lets the calling contract update a bundle, by passing in a list of tokens and a unique id.
+    function _updateBundle(Collateral[] memory _tokensToBind, uint256 _bundleId)
+        internal
+    {
+        require(_tokensToBind.length > 0, "!Tokens");
+
+        uint256 currentCount = bundle[_bundleId].count;
+        uint256 targetCount = _tokensToBind.length;
+        uint256 check = currentCount > targetCount ? currentCount : targetCount;
+
+        for (uint256 i = 0; i < check; i += 1) {
+            if (i < targetCount) {
+                _checkTokenType(_tokensToBind[i]);
+                bundle[_bundleId].collaterals[i] = _tokensToBind[i];
+            } else if (i < currentCount) {
+                delete bundle[_bundleId].collaterals[i];
+            }
+        }
+
+        bundle[_bundleId].count = targetCount;
+    }
+
+    /// @dev Lets the calling contract add a token to a bundle for a unique bundle id and index.
+    function _addTokenInBundle(
+        Collateral memory _tokenToBind,
+        uint256 _bundleId
+    ) internal {
+        _checkTokenType(_tokenToBind);
+        uint256 id = bundle[_bundleId].count;
+
+        bundle[_bundleId].collaterals[id] = _tokenToBind;
+        bundle[_bundleId].count += 1;
+    }
+
+    /// @dev Lets the calling contract update a token in a bundle for a unique bundle id and index.
+    function _updateTokenInBundle(
+        Collateral memory _tokenToBind,
+        uint256 _bundleId,
+        uint256 _index
+    ) internal {
+        require(_index < bundle[_bundleId].count, "index DNE");
+        _checkTokenType(_tokenToBind);
+        bundle[_bundleId].collaterals[_index] = _tokenToBind;
+    }
+
+    /// @dev Checks if the type of asset-contract is same as the TokenType specified.
+    function _checkTokenType(Collateral memory _token) 
+    internal view {
+        if (_token._collateralType == CollateralType.ERC721) {
+            try
+                IERC165(_token._collateralAddress).supportsInterface(0x80ac58cd)
+            returns (bool supported721) {
+                require(supported721, "TokenBundle: ERC721 Interface Not Supported");
+            } catch {
+                revert("TokenBundle: ERC721 Interface Not Supported");
+            }
+        } else if (_token._collateralType == CollateralType.ERC1155) {
+            try
+                IERC165(_token._collateralAddress).supportsInterface(0xd9b67a26)
+            returns (bool supported1155) {
+                require(supported1155, "TokenBundle: ERC1155 Interface Not Supported");
+            } catch {
```

Comment:
I'm worried that there are several tokens that don't implement the `supportsInterface` function that would cause this to fail

Reply:
This is just the way that thirdweb contract was written.  So we can either choose to change it or leave it as is based on this info. i agree w you.   I wonder why they did this.

Reply:
maybe we leave it for now and if we have any issues try to come up with a solution but not sure there is a good one

i.e. I wanted to figure out the token type to know which `transfer` function I have to call but this was the best I found
https://chat.openai.com/share/5fb9afbf-9d4b-4704-9359-145898ea906f